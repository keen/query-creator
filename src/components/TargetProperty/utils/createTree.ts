const setValue = (obj: Record<string, any>, path: string, value: string) => {
  const keys = path.split('.');
  let objectCopy = obj;

  while (keys.length - 1) {
    const key = keys.shift();
    if (!(key in objectCopy)) objectCopy[key] = {};
    objectCopy = objectCopy[key];
  }
  objectCopy[keys[0]] = [path, value];
};

export const createTree = (properties: Record<string, any>) =>
  Object.keys(properties)
    .sort()
    .reduce((acc, key) => {
      const propertyType = properties[key];
      setValue(acc, key, propertyType);
      return acc;
    }, {});
