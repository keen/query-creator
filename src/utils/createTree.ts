const setValue = (obj: Record<string, any>, path: string, value: string) => {
  const keys = path.split('.');
  let objectCopy = obj;

  while (keys.length - 1) {
    const key = keys.shift();
    if (!(key in objectCopy)) objectCopy[key] = {};
    objectCopy = objectCopy[key];
  }
  objectCopy[keys[0]] = [path, value];
};

/**
 * Creates tree from schema properties.
 *
 * @NOTES: reverse() is applied to support schema inconsistenty
 * and properties of list type
 *
 * @param properties - schema properties
 * @return tree object
 *
 */
export const createTree = (properties: Record<string, any>) =>
  Object.keys(properties)
    .sort()
    .reverse()
    .reduce((acc, key) => {
      const propertyType = properties[key];
      setValue(acc, key, propertyType);
      return acc;
    }, {});
