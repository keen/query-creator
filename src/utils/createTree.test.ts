import { createTree } from './createTree';

test('creates JSON tree', () => {
  const records = {
    'category.details.gender': 'male',
    'category.id': 'string',
    'category.details.name': 'string',
    age: 'number',
  };

  expect(createTree(records)).toMatchSnapshot();
});
