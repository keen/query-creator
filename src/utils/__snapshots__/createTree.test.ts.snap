// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates JSON tree 1`] = `
Object {
  "age": Array [
    "age",
    "number",
  ],
  "category": Object {
    "details": Object {
      "gender": Array [
        "category.details.gender",
        "male",
      ],
      "name": Array [
        "category.details.name",
        "string",
      ],
    },
    "id": Array [
      "category.id",
      "string",
    ],
  },
}
`;
